functions/Get-DbaCpuUsage.ps1

```diff
@@ -0,0 +1,144 @@
+		
+	.NOTES
+		Tags: CPU
+		dbatools PowerShell module (https://dbatools.io, clemaire@gmail.com)
+		Copyright (C) 2016 Chrissy LeMaire
+		License: GNU GPL v3 https://opensource.org/licenses/GPL-3.0
+
+	.LINK 
+		https://dbatools.io/Get-DbaCpuUsage
+
+	.EXAMPLE   
+		Get-DbaCpuUsage -SqlInstance sql2017
+		
+		Logs into the SQL Server instance "sql2017" and also the Computer itself (via WMI) to gather information
+		
+	.EXAMPLE 
+		$usage = Get-DbaCpuUsage -SqlInstance sql2017
+		$usage.Process
+		
+		Explores the processes (from Get-DbaProcess) associated with the usage results
+		
+	.EXAMPLE 
+		Get-DbaCpuUsage -SqlInstance sql2017 -SqlCredential (Get-Credential sqladmin) -Credential (Get-Credential ad\sqldba)
+		
+		Logs into the SQL instance using the SQL Login 'sqladmin' and then Windows instance as 'ad\sqldba'
+#>
+	[CmdletBinding()]
+	Param (
+		[parameter(Mandatory, ValueFromPipeline)]
+		[Alias("ServerInstance", "SqlServer", "SqlServers")]
+		[DbaInstanceParameter[]]$SqlInstance,
+		[PSCredential]$SqlCredential,
+		[PSCredential]$Credential,
+		[int]$Threshold = 0,
+		[switch]$EnableException
+	)
+	begin {
+		# This can likely be enumerated but I don't know hows
+		$threadstates = [pscustomobject]@{
+			0 = 'Initialized. It is recognized by the microkernel.'
+			1 = 'Ready. It is prepared to run on the next available processor.'
+			2 = 'Running. It is executing.'
+			3 = 'Standby. It is about to run. Only one thread may be in this state at a time.'
+			4 = 'Terminated. It is finished executing.'
+			5 = 'Waiting. It is not ready for the processor. When ready, it will be rescheduled.'
+			6 = 'Transition. The thread is waiting for resources other than the processor.'
+			7 = 'Unknown. The thread state is unknown.'
+		}
+		
+		$threadwaitreasons = [pscustomobject]@{
+			0 = 'Executive'
+			1 = 'FreePage'
+			2 = 'PageIn'
+			3 = 'PoolAllocation'
+			4 = 'ExecutionDelay'
+			5 = 'FreePage'
+			6 = 'PageIn'
+			7 = 'Executive'
+			8 = 'FreePage'
+			9 = 'PageIn'
+			10 = 'PoolAllocation'
+			11 = 'ExecutionDelay'
+			12 = 'FreePage'
+			13 = 'PageIn'
+			14 = 'EventPairHigh'
+			15 = 'EventPairLow'
+			16 = 'LPCReceive'
+			17 = 'LPCReply'
+			18 = 'VirtualMemory'
+			19 = 'PageOut'
+			20 = 'Unknown'
+		}
+	}
+	process {
+		foreach ($instance in $SqlInstance) {
+			try {
+				Write-Message -Level Verbose -Message "Connecting to $instance"
+				$server = Connect-SqlInstance -SqlInstance $instance -SqlCredential $sqlcredential
+			}
+			catch {
+				Stop-Function -Message "Failure" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
+			}
+			
+			$processes = Get-DbaProcess -SqlInstance $server
+			$threads = Get-DbaCmObject -ComputerName $instance.ComputerName -ClassName Win32_PerfFormattedData_PerfProc_Thread -Credential $Credential | Where-Object { $_.Name -like 'sql*' -and $_.PercentProcessorTime -ge $Threshold }
+			$spidcollection = $server.Query("select spid, kpid from sysprocesses")
```

Comment:
I would disagree with using a view that is documented to be on the depreciation list. This should only be used to cover older versions of SQL Server.

Reply:
i saw it was deprecated but i'm not convinced they'll actually remove it ;)

Reply:
It won't go away because Activity Monitor is SSMS utilizes it. However, what is wrong with using current views that give the same information?

Reply:
It is not really the point that it is not going away but we are using views on the depreciation list, which by Microsoft terms is a bad practice. First thing folks see is that we are using SQL Server 2000 views.

Something like this should provide you the same information...

```sql
select s.host_name, s.host_process_id, t.os_thread_id, s.session_id
from sys.dm_exec_sessions s
   join sys.dm_exec_requests er on s.session_id = er.session_id
   join sys.dm_os_workers w on er.task_address = w.task_address
   join sys.dm_os_threads t on w.thread_address = t.thread_address;
```